core/src/main/java/com/linecorp/armeria/client/proxy/ProxyConfig.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.proxy;
+
+import static com.linecorp.armeria.client.proxy.NoopProxyConfig.NOOP_PROXY_CONFIG;
+import static java.util.Objects.requireNonNull;
+
+import java.net.InetSocketAddress;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.ClientFactory;
+
+/**
+ * Contains base configuration for proxy related settings used in {@link ClientFactory}.
+ */
+public abstract class ProxyConfig {
+
+    /**
+     * Creates a {@code ProxyConfig} configuration for SOCKS4 protocol.
+     * @param proxyAddress The proxy address.
+     */
+    public static Socks4ProxyConfigBuilder socks4(InetSocketAddress proxyAddress) {
+        return new Socks4ProxyConfigBuilder(requireNonNull(proxyAddress));
```

Comment:
I think Armeria would not use the following API design. You might want to start with `builder()...` to build `ProxyConfig` pattern. 
```java
ProxyConfig.builder()
           .socks4(...)
           .username(...)
           .build()
```


Or How about making factory methods as we did in `Backoff`?
https://github.com/line/armeria/blob/bd28eb2499c9149613098de17bc1479919dc3fd7/core/src/main/java/com/linecorp/armeria/client/retry/Backoff.java#L50-L62

Reply:
How about `builderForSocks4`, etc?

Reply:
Alternatively, we could give up builder API and just add various factory methods:

```java
public static Socks4ProxyConfig socks4(address);
public static Socks4ProxyConfig socks4(address, username);
public static Socks5ProxyConfig socks5(address);
public static Socks5ProxyConfig socks5(address, username, password);
public static ConnectProxyConfig connect(address);
public static ConnectProxyConfig connect(address, HttpHeaders?);
```

Reply:
> start with builder()
It may be hard to express  the idea that only one of the proxy protocols can be selected with this. As @trustin suggested, we would have to have sth like `builderForSocks4` if we want to go this route.

> give up builder API and just add various factory methods
I initially gave up on this idea bc I don't like nullable parameters, but I do feel like this pattern may be slightly overkill.

Decided to go with factory methods to save users from typing the extra `build()` 😉